### ISSUE TITLE
Drift Correction issues with multi-shank probe

### ISSUE THREAD

**Issue by Selmaan** [MouseLand/Kilosort#686]
### Describe the issue:

I am using kilosort 4 with a 2-shank 64ch probe. Most of my recordings have essentially no drift and the algorithm has been working beautifully, thanks for the work put into this! However I am concerned there is a bug with how depth is handled for multi-shank probes for the drift correction step, which screws up the drift correction module. 

Below is the channel map displayed in the ks4 gui. Note that the probe goes from 0-330um depth.
![image](https://github.com/MouseLand/Kilosort/assets/5429274/2527532b-5a55-4e05-a3a4-e9fcdc0079bf) 
![image](https://github.com/MouseLand/Kilosort/assets/5429274/98249bcd-23a0-4703-914a-ff43eeec351f)

this is the spike position across probe colored by cluster, showing that spikes are well localized on the probe and match the channel map dimensions
![image](https://github.com/MouseLand/Kilosort/assets/5429274/74a3ab77-b41b-4792-b1ab-b0b4e4083661)

However when I look at the spike map used in drift correction, I see that spike depths range twice what they should, i.e. from 0 to ~600
![image](https://github.com/MouseLand/Kilosort/assets/5429274/ce962964-ebe1-42ed-8f04-c0c5b4dfec01)

Furthermore, even though there is a clear single instability in the recording corresponding to a shift of <20um, the drift correction algorithm has a noisy and unstable estimate which vastly overcompensates, flickering btw 0 and 60ums of drift early in the recording:
![image](https://github.com/MouseLand/Kilosort/assets/5429274/1c81e5f8-abb2-444a-abb3-8fc14233d6e2)

(this is using 180,000 batch size with a 30,000 sampling rate, to account for fewer spikes on a 64ch probe than a neuropixels)

### Version information:

Following install instructions, using python 3.9, pytorch-cuda 11.8, kilosort v4.0.6

**Comment by jacobpennington** [comment]
The issue with the drift scatter plot is a bug in the plot code, I'll have that fixed soon. Thanks for catching that!
As for the noisy drift estimate, I would take that as a sign that you should not use drift correction for that probe (i.e. set `nblocks = 0`). ~64 channels (on a single shank) is the minimum we recommend to get reliable drift estimates.

**Comment by Selmaan** [comment]
Is there any way to enforce a smoother drift estimate (besides using massive batch sizes I guess)? I am working in an unusual model species and brain area where we have many more units per unit volume than typical for mouse cortex (usually 100-200 units per recording), and I know from previous experience this kind of data is often sufficient to get reliable drift estimates (at least assuming rigidity, which seems valid to me on visual examination). But occasionally there are units which essentially turn on or off throughout the recording and disrupt motion estimates. I think you can see some of this in the spike amplitude plot I shared? I often end up turning drift correction off entirely, because estimated drift is usually +-10um or so, unless occasionally I get these massive errors. But it would be fantastic to impose some kind of smoother or robustness or prior on the drift correction...

**Comment by marius10p** [comment]
There is already a temporal smoothing step in the drift correction, which smooths the "Z" cross-correlation curves before taking the max (which is the drift estimate). This effectively should work like a prior. We could make the smoothing constant a parameter (@jacobpennington). 

However, Jacob and I discussed this case today and we think there might be more going on. It looks like there is a bit of a large physical readjustment happening around 3000 sec. There are many neurons that are either only active before or only active after the re-adjustment. This could be due to physical modifications in tissue or distortions beyond what we can fix with drift correction. In the extreme case, it would correspond to neurons dying off or getting silenced by the big physical movement. 

If you increase the batch size and it still does not help, then I think my hypothesis above might be true. In that case, the next best thing you could is split data around such critical points, spike sort separately, and use one of the packages that match neurons over days to make the matchings, since that matching is likely to be highly nonlinear.

**Comment by Selmaan** [comment]
Let me get back to you with a better example, I agree with your point about
the plot I shared in the first post and I have better examples to
demonstrate the issue I mentioned in the most recent post. But the short
version is I would definitely explore adjusting the smoothing if you can
expose that to the user in some way.

**Comment by Selmaan** [comment]
To be clear, it sounds like the original bug I mentioned is just a plotting issue and will be resolved, so we can close this issue if you'd prefer. But on the topic of smoothing out drift correction estimates, this is an example of a spike position plot for a cleaner recording. You can see some high amplitude neurons simply turn on or off, and reappear at the same location after turning off, and this is not particularly synchronized across neurons. Overall the recording is pretty stable, although there is a bit of drift particularly early in the recording. My experience is that this kind of data sometimes catastrophically fails the drift correct step, and so I usually turn drift correction off entirely, although that's clearly not optimal
<img width="1750" alt="image" src="https://github.com/MouseLand/Kilosort/assets/5429274/021f63d9-1078-4c56-bc4c-ce1a20386020">

same data, slightly different plot:
<img width="1874" alt="image" src="https://github.com/MouseLand/Kilosort/assets/5429274/9ac2e0d7-e736-42f6-ab31-a864f63ef1a5">

**Comment by jacobpennington** [comment]
@marius10p is `sig_interp` not the smoothness parameter you're referring to?
```
    'sig_interp': {
        'gui_name': 'sig_interp', 'type': float, 'min': 0, 'max': np.inf,
        'exclude': [0], 'default': 20, 'step': 'preprocessing',
        'description':
            """
            For drift correction, sigma for interpolation (spatial standard
            deviation). Approximate smoothness scale in units of microns.
            """
    },
```

**Comment by jacobpennington** [comment]
@Selmaan I'm going to add the smoothness parameter Marius mentioned, it's not the one I mentioned above. In the meantime, if you want to try it with your data and see if it helps you can change the values on line 139 in `kilosort.datashift.py`:
```
    # smooth the dot-product matrices across blocks, batches and across vertical offsets
    dcs = gaussian_filter(dcs, [0.5, 0.5, 0.5])
```
The 0.5s are the sigma for x,y, and time, respectively.

**Comment by jacobpennington** [comment]
This parameter has been added in the latest version as `drift_smoothing`. I'll close this as resolved for now, but if you run into any more problems please let us know.

**Comment by jacobpennington** [comment]
```
Amount of gaussian smoothing to apply to the spatiotemporal drift
estimation, for x,y,time axes in units of registration blocks
(for x,y axes) and batch size (for time axis). The x,y smoothing has
no effect for `nblocks = 1`.
```

**Comment by Selmaan** [comment]
Ah I see. Just FYI, I ran this with t=5-block smoothing for the following example session with a pretty stable recording. Oddly, the resulting drift trace appears to have unsmoothed noise...is smoothing applied upstream or downstream of this plot?
![image](https://github.com/MouseLand/Kilosort/assets/5429274/945f9372-181a-467a-8b06-0d5e77a94df8)
![image](https://github.com/MouseLand/Kilosort/assets/5429274/be3cfd85-e592-40ea-affd-7a6f15a59bd1)

**Comment by Selmaan** [comment]
ok. I'll look into this more then trying a couple different settings...my expectation was that there would be less 'spikiness' in the plots

**Comment by marius10p** [comment]
@Selmaan that is a very stable drift trace, and the apparent "noise" is just due to hitting the integer floor of the estimation process which is about half a micron for NP. You won't have drift related problems with this data (and you could even turn drift correction off). This is very different from the first example  you posted.

**Comment by Selmaan** [comment]
Yes it is, and I understand now about the discretization causing apparent "noise", so it was probably a bad place to start. Here are results for a session with noticeable drift, and a high temporal smoothing param (10). It is possibly oversmoothed and I could slightly reduce the smoothing param, but it's definitely tracking the coarse drift. There appears to be some kind of edge effect on the last bin? 
![image](https://github.com/MouseLand/Kilosort/assets/5429274/8d984bae-8bdf-4f24-9874-6c47bb9650bc)
![image](https://github.com/MouseLand/Kilosort/assets/5429274/e411941b-4afd-4d69-bff0-4bbcf1b22336)

**Comment by marius10p** [comment]
Does that correspond to batches with no spikes, or very few spikes? Do you wait for the probe to settle at the beginning of the recording?

**Comment by Selmaan** [comment]
Sorry, I'm not sure the best way to determine that, although it doesn't seem to me like these times are missing spikes based on a very coarse judgment in the drift scatter. More generally this level of drift is unusual, I experienced it often recording from one bird for unclear reasons, but doing reasonable things (e.g. advance probe past target recording depth, retract, and leave to settle before recording) leads to much less drift in most of my recordings, like the more stable traces I've shared. Overall I'm usually able to manage drift mostly on the data collection side which is of course preferable, and usually drift correction works well. So here I was just focusing on a particularly challenging case, and also trying to ensure that the new parameter you added for smoothing drift correction is working as expected. On that last point, I'm still a bit unsure, or at least results are not matching my intuition. 

For example, here is a drift map for the initial session I shared in this thread, (which we all agreed is problematic for a couple reasons, and now looks different with the correct depth plotting for multi-shank probes). 
![image](https://github.com/MouseLand/Kilosort/assets/5429274/a01f94e6-0b11-460b-81ec-e75cba538111)

The only thing I wanted to highlight is that I seem to be getting quite similar estimated drift traces regardless of the smoothing parameter. So for example here is the estimated drift trace when massively increasing the temporal smoothing (7s batches, and [0.5, 0.5, 60.0] drift smoothing)
![image](https://github.com/MouseLand/Kilosort/assets/5429274/4c0c3839-5b30-4058-b13f-07a07bceac2d)
and zoom in
![image](https://github.com/MouseLand/Kilosort/assets/5429274/4e8057a2-e7d1-4b7e-bfe3-9cdb346e618d)

all that being said, the original issue of depth for multishank probes being misplotted is definitely solved, and it is rare for my data to have this level of drift issue, so I am ok closing this if you think drift correction is working fine and this is just a data issue!

**Comment by jacobpennington** [comment]
I'll check this on some of our high drift data as well to be sure, but FYI the documentation for the parameter was incorrect. The first axis is for a correlation we're computing (don't recommend changing this value), the second is for time (in units of batches), and the third is for the vertical axis on the probe (units of registration blocks).

**Comment by jacobpennington** [comment]
This is what I got when testing on a high drift simulation for different values for the second and third (time in batches, y in registration blocks) positions. Let me know if anything is unclear.

![smoothing_test](https://github.com/MouseLand/Kilosort/assets/35777376/0eb10111-b0e0-4b5d-aa08-9cb0f458c862)

**Comment by Selmaan** [comment]
I am a little unclear actually. The registration block smoothing makes
complete sense to me. But the temporal smoothing by batch is a bit
confusing, it looks here like the drift traces are more coarsely
discretized? I expected it to look smoother and/or to dampen fluctuations
as the temporal smoothing gets very high. So maybe I'm not getting what the
parameter really does?

**Comment by jacobpennington** [comment]
You're right, that is counter-intuitive. Sorry, I was thinking about it backwards. I did some more testing and zoomed in a bit for a single block. It looks like increasing the time smoothing *does* smooth out the trace as expected, but also results in outliers for some points which is why those zoomed out traces look less smooth overall. I'll look into this some more to figure out why that's happening.
![image](https://github.com/MouseLand/Kilosort/assets/35777376/bde4f7ca-9e00-4809-91d9-7d7401f943f1)

**Comment by jacobpennington** [comment]
Okay, we did identify a bug related to this that we'll have to work on. Thanks again for your patience and for helping point that out.

**Comment by Selmaan** [comment]
no problem, thanks for working on this feature!

**Comment by cgallimore25** [comment]
> " In that case, the next best thing you could is split data around such critical points, spike sort separately, and use one of the packages that match neurons over days to make the matchings, since that matching is likely to be highly nonlinear."

@marius10p I'm finding myself in a situation where it may be best for me to sort 2 experimental runs separately, as they were done on a low channel count probe (16 channels, 50um inter-contact distance) and separated by a 30-min wait period (pre- and post- drug application).  These multielectrode specs make drift correction inaccurate.  Would you mind pointing me to one of these packages you mention that matches neurons over days?  

I'm trying to brainstorm a good way to match the same neuron across the 2 recordings which has drifted 1 channel above/below its original position in the baseline recording.  It helps that in my recordings, individual neurons generally drift no more than 1 channel (making knowledge of its channel index very helpful), though a quantitative way would be best. Thanks

**Comment by marius10p** [comment]
Sure, here's two that I know of: 

https://github.com/EnnyvanBeest/UnitMatch
https://github.com/janelia-TDHarrisLab/Yuan-Neuron_Tracking
